ml4gw/transforms/qtransform.py

```diff
@@ -0,0 +1,444 @@
+        frange: List[float] = [0, torch.inf],
+        mismatch: float = 0.2,
+    ):
+        super().__init__()
+        self.qrange = qrange
+        self.mismatch = mismatch
+        self.qs = self.get_qs()
+        self.frange = frange
+
+        # Deliberately doing something different from GWpy here.
+        # Their final frange is the intersection of the frange
+        # from each q. This implementation uses the frange of
+        # the chosen q.
+        self.q_transforms = torch.nn.ModuleList(
+            [
+                SingleQTransform(
+                    duration=duration,
+                    sample_rate=sample_rate,
+                    q=q,
+                    frange=self.frange.copy(),
+                    mismatch=self.mismatch,
+                )
+                for q in self.qs
+            ]
+        )
+
+    def get_qs(self):
+        """
+        Determine the values of Q to try for the set of Q-transforms
+        """
+        deltam = 2 * (self.mismatch / 3.0) ** (1 / 2.0)
+        cumum = math.log(self.qrange[1] / self.qrange[0]) / 2 ** (1 / 2.0)
+        nplanes = int(max(math.ceil(cumum / deltam), 1))
+        dq = cumum / nplanes
+        qs = [
+            self.qrange[0] * math.exp(2 ** (1 / 2.0) * dq * (i + 0.5))
+            for i in range(nplanes)
+        ]
+        return qs
+
+    def forward(
+        self,
+        X: torch.Tensor,
+        num_f_bins: int,
+        num_t_bins: int,
+        fsearch_range: List[float] = None,
+        norm: str = "median",
+    ):
+        """
+        Compute the set of QTiles for each Q transform and determine which
+        has the highest energy value. Interpolate and return the
+        corresponding set of tiles.
+
+        Args:
+            X:
+                Time series of data. Should have the duration and sample rate
+                used to initialize this object. Expected input shape is
+                `(B, C, T)`, where T is the number of samples, C is the number
+                of channels, and B is the number of batches. If less than
+                three-dimensional, axes will be added during Q-tile
+                computation.
+            num_f_bins:
+                Number of frequency bins to interpolate to. Because the
+                frequency spacing of the Q-tiles is in log-space, the frequency
+                interpolation is log-spaced as well.
+            num_t_bins:
+                Number of time bins to interpolate to.
+            fsearch_range:
+                The lower and upper frequency values within which to search
+                for the maximum energy
+            norm:
+                The method of interpolation used by each QTile
+
+        Returns:
+            An interpolated Q-transform for the batch of data. Output will
+            have one more dimension than the input
+        """
+        for transform in self.q_transforms:
+            transform.compute_qtiles(X, norm)
```

Comment:
Continuing above discussion: here is where my confusion stemmed from. Was unclear what was data was being manipulated / adjusted to do the transform. 

Won't hold this up if you think this is the best way to do it.

Reply:
Okay, cool. I'll merge this in, but keep this in mind for future updates